Add docs regaring --to-remote option for add/import-url (#2091)
iterative/dvc.org: content/docs/command-reference/import-url.md

# import-url

Download a file or directory from a supported URL (for example `s3://`,

> `ssh://`, and other protocols) into the <abbr>workspace</abbr> (or to the

Reviewer: I think we need to work on this a bit more ... probably get rid of Download? 

brackets look too complicated for the very beginning of the doc

cc @jorgeorpinel

Reviewer 2: Agree. "Download to remote storage" doesn't seem clear... 🤔 

> The parenthesis here look OK to me though. and that's already in https://dvc.org/doc/command-reference/import-url (out of scope to change here?)

Reviewer 2: But OK, if we completely rewrite this anyway, we can address the parenthesis situation along the way. Maybe something like

``Track a file or directory found in an external location (`s3://`, `/local/path`, etc.), and download it to the local project, or make a copy in [remote storage](/doc/command-reference/remote).``

Reviewer 2: Please notice similar suggestion (for add) in https://github.com/iterative/dvc/pull/5198#pullrequestreview-566897428

Reviewer 2: p.s. but I guess let's decide here first, where the description can be a little longer, and then we can update similarly in core?

> `ssh://`, and other protocols) into either the <abbr>workspace</abbr> or to a

Reviewer: we still create .dvc file in this case, right? I would not put this special way of importing into this synopsis part

Reviewer: could you please briefly explain why do we need to support `--to-remote` in both - `dvc add` and `dvc import-url`? Just want to be sure I understand this. Why `dvc add` is not enough?

Reviewer 2: I have the same Q. Maybe iterative/dvc/issues/4527 is a better route for imports?

Author: > could you please briefly explain why do we need to support --to-remote in both - dvc add and dvc import-url?

Initially it was only in the `import-url`, though for users who don't want to track the original location we went ahead and also applied the same thing for the `add`.

Reviewer 2: Hmmm. I thought `add` was the original/main command here, per iterative/dvc/issues/4520. Anyway, I def. see how it makes sense as an "alternative" to `add --external` and thinking about it, iterative/dvc/issues/4527 is not enough (to transfer to remote directly) in the case of `import-url`.

What about plain `import` though? Shouldn't it be included as well? Basically to move specific data from one remote to another, while at the same time tracking it independently in a new project (a workaround to [chained imports](https://github.com/iterative/dvc/issues/3305)).

```usage
usage: dvc import-url [-h] [-q | -v] [--file <filename>] [--no-exec]

>                       [--to-remote] [-r <name>] [-j <number>]
>                       [--desc <text>]

Reviewer 2: Same (https://github.com/iterative/dvc/pull/5198#discussion_r556712627):
```suggestion
                      [--to-remote] [-j <number>] [--desc <text>]
```

Reviewer 2: I guess we can skip this one too for now.

## Description

In some cases it's convenient to add a data file or directory from an external

> location into the workspace (or to the
> [remote storage](/doc/command-reference/remote), such that it can be updated

Reviewer 2: Typos
```suggestion
location into the workspace (or to
[remote storage](/doc/command-reference/remote)), such that it can be updated
```

having to manually copy files from the supported locations (listed below), which
may require installing a different tool for each type.

> When you don't actually want to store the whole data file / directory in your
> local workspace but rather import it directly to the remote storage,
> `--to-remote` option can be given. See the
> ["import straight to remote"](#example-import-straight-to-the-remote) example
> for more details.

Reviewer 2: Copy edit:
```suggestion
When you don't want to store the target data in your local system, you can still
create an import `.dvc` file while transferring a file or directory directly to
remote storage, by using the `--to-remote` option. See the
[Import straight to remote](#example-import-straight-to-remote) example for
more details.
```

Reviewer 2: But in `add` we created a whole section, while in this ref. it's one new paragraph in the description + an example. Which option is better (simple=better)? Should we be consistent in both refs?

Reviewer 2: After reviewing the example in `import-url` I think the change we did in that ref. makes more sense: just a paragraph in the cmd Description, and link to an example that explains more visually.

Author: So should I add the same example to the `add` and remove the section @jorgeorpinel?

Reviewer 2: Yes, I now like the example option better, so there is space to explain the caveats there (and just add a summary paragraph about it in the description). Either way it could work as long as the changes in add and import-url are consistent. And it's OK if the examples are very similar, yes. Thanks

Reviewer 2: Hi! I left a few last comments in the import-url Example so once that's done we can copy (adapt) it to add as well (:

> [Import straight to remote](#example-transfer-to-remote-storage) example for

Reviewer 2: ```suggestion
[Transfer to remote storage](#example-transfer-to-remote-storage) example for
```

> [Import straight to remote](#example-import-straight-to-remote) example for more

Reviewer 2: ```suggestion
[Import straight to remote](#example-transfer-to-remote-storage) example for more
```

> more details.

Reviewer 2: Can we put something like this in the `add` reference Description (instead of the **Transferring data directly to remote storage** section) ?

  finish the operation(s)); or if the target data already exist locally and you
  want to "DVCfy" this state of the project (see also `dvc commit`).

- `--to-remote` - Import an external target, but don't move it into the

>   workspace, nor cache it. [Transfer](#example-import-straight-to-the-remote) it

Reviewer 2: ```suggestion
  workspace, nor cache it. [Transfer](#example-transfer-to-remote-storage) it
```
May need wrapping too

>   directly to remote storage instead. Use `dvc pull` to get the data locally.

Reviewer 2: We should probably mention that the default remote is used if set, and otherwise you need to provide -r.

Reviewer 2: Same in add

- `-r <name>`, `--remote <name>` - name of the

>   [remote storage](/doc/command-reference/remote)

Reviewer 2: Let's mention this is specifically for --to-remote please

  finish the operation(s)); or if the target data already exist locally and you
  want to "DVCfy" this state of the project (see also `dvc commit`).

> - `--to-remote` - Import an external target, but don't move it into the

Reviewer 2: ```suggestion
- `--to-remote` - import an external target, but don't move it into the
```

  already exist locally and you want to "DVCfy" this state of the project (see
  also `dvc commit`).

- `--to-remote` - import data straight to remote storage and create a .dvc file.
  Check [this](#example-import-straight-to-the-remote) section for the details.

- `-r <name>`, `--remote <name>` - name of the

>   [remote storage](/doc/command-reference/remote)

Reviewer 2: Same (merge to previous option).

Reviewer 2: Same (skip for now)

Running stage 'prepare' with command:
	python src/prepare.py data/data.xml
```

## Example: Transfer to remote storage

When you have a large dataset in an external location, you may want to import it
to you project without downloading it to the local file system (for using it
later/elsewhere). The `--to-remote` option lets you skip the download, while
storing the imported data [remotely](/doc/command-reference/remote). Let's
initialize a DVC project, and setup a remote:

```dvc
$ mkdir example # workspace
$ mkdir /tmp/dvc-storage
$ cd example
$ git init
$ dvc init
$ dvc remote add myremote /tmp/dvc-storage
```

Now let's create an import `.dvc` file without downloading the target data,
transferring it directly to remote storage instead:

```
$ dvc import-url https://data.dvc.org/get-started/data.xml data.xml \
                 --to-remote -r local_remote
...

> ```
> 
> When you run the `import-url` with `--to-remote`, you pass as usual the remote
> location and the output filename, afterward if you haven't set a default
> [remote](/doc/command-reference/remote) yet, you can simply pass the name of the
> remote with `-r`/`--remote` flag and it will start the transfer and leave a DVC
> file as an only side effect on your workspace (everything else happens in the
> remote storage unit)

Reviewer 2: ````suggestion
```
````

After the previous changes we did I think all of this is redundant, and the default remote stuff can be explained in the Options section.

Reviewer 2: BUT I think we could extend the code block (or add a short sentence and another code block) to show what happened in the workspace (only .dvc file created, nothing cached or linked) and in the remote (imported data is stored there).

Running stage 'prepare' with command:
	python src/prepare.py data/data.xml
```

## Example: Transfer to remote storage

> When you have a large dataset in an external location and want to import it to
> you project without using your local disk, you can use the `--to-remote` option
> to transfer it directly to remote storage. Let's initialize a new project, and
> add a local [remote](/doc/command-reference/remote):

Reviewer 2: ```suggestion
When you have a large dataset in an external location, you may want to import it
to you project without downloading it to the local file system. The
`--to-remote` option lets you transfer it directly to
[remote storage](/doc/command-reference/remote). Let's initialize a DVC project,
and setup a remote:
```

Running stage 'prepare' with command:
	python src/prepare.py data/data.xml
```

## Example: Transfer to remote storage

> When you have a large dataset in an external location, you may want to import it
> to you project without downloading it to the local file system. The
> `--to-remote` option lets you transfer it directly to
> [remote storage](/doc/command-reference/remote). Let's initialize a DVC project,
> and setup a remote:

Reviewer 2: ```suggestion
When you have a large dataset in an external location, you may want to import it
to you project without downloading it to the local file system (for using it
later/elsewhere). The `--to-remote` option lets you skip the download, while
storing the imported data [remotely](/doc/command-reference/remote). Let's
initialize a DVC project, and setup a remote:
```

```dvc
$ mkdir example # workspace

> $ mkdir /tmp/dvc-storage # remote storage
> $ cd example
> $ git init
> $ dvc init
> $ dvc remote add local_remote /tmp/dvc-storage

Reviewer 2: I think I'll get rid of the "local remote" term here to avoid confusions
```suggestion
$ mkdir /tmp/dvc-storage
$ cd example
$ git init
$ dvc init
$ dvc remote add myremote /tmp/dvc-storage
```

```

> Now let's create an import .dvc file without downloading the target data, but
> transferring directly to remote storage instead:

Reviewer 2: ```suggestion
Now let's create an import `.dvc` file without downloading the target data,
transferring it directly to remote storage instead:
```

```
$ dvc import-url https://data.dvc.org/get-started/data.xml data.xml \

>                  --to-remote -r local_remote

Reviewer 2: ```suggestion
                 --to-remote -r myremote
```

> To track the changes with git, run:
> 
>         git add data.xml.dvc

Reviewer 2: Prob no need for this output. I'll remove it:
```suggestion
...
```

Running stage 'prepare' with command:
	python src/prepare.py data/data.xml
```

> ## Example: Import straight to the remote

Reviewer 2: ```suggestion
## Example: Import straight to remote
```

Reviewer 2: But I think we should stick with the term "transfer" here, not sure.
```suggestion
## Example: Transfer to remote storage
```
> If we change this let's remember to update the anchored link in the description.

When you have a massive dataset in a distant location, and working on a computer
which can't actually store it locally (due to not having enough disk space) but

> you still want to take it under control of DVC just like in the scenario of

Reviewer 2: Good info. but this is repetitive (already explained in the description where this example is linked to).

Let's jump right into the example instead? Well, we can leave a 1-2 sentence intro if needed.

importing it and then pushing it to the remote, then you can use `--to-remote`
flag.

It will try to import the data into the remote storage that you choose, and when
you or any of your colleagues want to copy the data to their systems, they could
just simply [pull](/doc/command-reference/remote). Let's do a simple example

> We initalize 2 directories, one being the remote storage unit and the other one
> is the workspace.
> 
> ```dvc
> $ mkdir /tmp/dvc-import-url-straight-to-remote/
> $ mkdir /tmp/remote
> $ cd /tmp/dvc-import-url-straight-to-remote/
> $ git init
> $ dvc init
> $ dvc remote add tmp_remote /tmp/remote

Reviewer 2: I'd basically start here, with something like:

````suggestion
Let's initialize a new project, and add a local [remote](/doc/command-reference/remote):

```dvc
$ mkdir example # workspace
$ mkdir /tmp/dvc-storage # remote storage
$ cd example
$ git init
$ dvc init
$ dvc remote add local_remote /tmp/dvc-storage
````
> ^ I suggested some changes to the example for readability 🙂

```

For transferring a source from a remote location, to the given remote you can
combine `import-url` with `--to-remote` option which basically does the whole

> importing and [push](/doc/command-reference/push)ing operation under the hood

Reviewer 2: Again, very accurate, but it seems like too much text, probably. We can just quickly explain the next step. Something like:

> Now let's create an import .dvc file without downloading the target data, but transferring directly to remote storage instead:

And that's it, onto the next code block ⬇️

Reviewer 2: After the code block where `dvc import-url` is used, THEN we can explain what we did 🙂 (explain the option and the argument sent to it).

but without actually downloading everything in once, but rather transferring
gradually.

When you run the `import-url` with `--to-remote`, you pass as usual the remote
location and the output filename, afterward if you haven't set a default
[remote](/doc/command-reference/remote) yet, you can simply pass the name of the
remote with `-r`/`--remote` flag and it will start the transfer and leave a DVC
file as an only side effect on your workspace (everything else happens in the
remote storage unit)

```

> $ dvc import-url https://data.dvc.org/get-started/data.xml data.xml --to-remote -r tmp_remote

Reviewer 2: Please break code blocks at max. 72 chars:
```suggestion
$ dvc import-url https://data.dvc.org/get-started/data.xml data.xml \
                 --to-remote local_remote
```
> Plus updates to match my previous suggestions

To track the changes with git, run:

        git add data.xml.dvc
```

> Whenever anyone wants to actually get this file, like when they have a system
> which can handle it, it is just a simple [pull](/doc/command-reference/pull)
> operation.
> 
> ```
>  $ dvc pull data.xml.dvc -r tmp_remote
> 

Reviewer 2: Copy edit
````suggestion
Whenever anyone wants to actually download the imported data (for example from a
system that can handle it), they can use `dvc pull` as usual:

```
 $ dvc pull data.xml.dvc -r local_remote
````

Running stage 'prepare' with command:
	python src/prepare.py data/data.xml
```

## Example: Transfer to remote storage

When you have a large dataset in an external location, you may want to import it
to you project without downloading it to the local file system (for using it

> later/elsewhere). The `--to-remote` option lets you skip the download, while

Reviewer 2: And copy over the Example as well (will need some some adapting). Thanks

storing the imported data [remotely](/doc/command-reference/remote). Let's
initialize a DVC project, and setup a remote:

```dvc

> $ mkdir example # workspace
> $ mkdir /tmp/dvc-storage
> $ cd example
> $ git init
> $ dvc init
> $ dvc remote add myremote /tmp/dvc-storage

Reviewer 2: ```suggestion
$ mkdir example # workspace
$ cd example
$ git init
$ dvc init
$ mkdir /tmp/dvc-storage
$ dvc remote add myremote /tmp/dvc-storage
```

```

Now let's create an import `.dvc` file without downloading the target data,
transferring it directly to remote storage instead:

> ```

Reviewer 2: ````suggestion
```dvc
````

$ dvc import-url https://data.dvc.org/get-started/data.xml data.xml \
                 --to-remote -r myremote
...
```

> After importing `data.xml` to our remote storage unit, the only change in our
> local workspace is the newly created dvc file for `data.xml`.
> 
> ```
> $ ls
> data.xml.rc

Reviewer 2: ````suggestion
The only change in our local <abbr>workspace</abbr> is a newly created import
`.dvc` file:

```dvc
$ ls
data.xml.dvc
````
